hive/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java

```diff
@@ -0,0 +1,238 @@
+import java.io.IOException;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.Transaction;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.types.Types;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.PartitionSpec.builderFor;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+public class HiveCreateReplaceTableTest extends HiveMetastoreTest {
+
+  private static final String TABLE_NAME = "tbl";
+  private static final TableIdentifier TABLE_IDENTIFIER = TableIdentifier.of(DB_NAME, TABLE_NAME);
+  private static final Schema SCHEMA = new Schema(
+      required(3, "id", Types.IntegerType.get()),
+      required(4, "data", Types.StringType.get())
+  );
+  private static final PartitionSpec SPEC = builderFor(SCHEMA)
+      .identity("id")
+      .build();
+
+  @Rule
+  public ExpectedException exceptionRule = ExpectedException.none();
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private String tableLocation;
+
+  @Before
+  public void createTableLocation() throws IOException {
+    tableLocation = temp.newFolder("hive-").getPath();
+  }
+
+  @After
+  public void cleanup() {
+    catalog.dropTable(TABLE_IDENTIFIER);
+  }
+
+  @Test
+  public void testCreateTableTxn() {
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    txn.updateProperties()
+        .set("prop", "value")
+        .commit();
+
+    // verify the table is still not visible before the transaction is committed
+    Assert.assertFalse(catalog.tableExists(TABLE_IDENTIFIER));
+
+    txn.commitTransaction();
+
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    Assert.assertEquals("Table props should match", "value", table.properties().get("prop"));
+  }
+
+  @Test
+  public void testCreateTableTxnTableCreatedConcurrently() {
+    exceptionRule.expect(RuntimeException.class);
+    exceptionRule.expectMessage("Metastore operation failed");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+
+    // create the table concurrently
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // expect the transaction to fail
+    txn.commitTransaction();
```

Comment:
I prefer to use `TestHelpers.assertThrows` for cases like this instead of the `ExpectedException` rule because I think what is happening is more clear in the test and so that you can make more assertions after the exception is thrown to validate that the conflicting table has not changed.

Reply:
This is minor, so don't worry about it if it would be a lot of work to convert.

Reply:
Let me consume the test jar for iceberg-api in iceberg-hive then.

Reply:
Wait, what is the difference between `AssertHelpers` and `TestHelpers`?

Reply:
`AssertHelpers` is the one in core, so probably use that one.

Before we were producing test Jars to share code, we copied this in a few places. Should probably clean that up.